Major Refactoring

splitbrain opened the pull request:
This is a major refactoring of the plugin's code base with the goal to make it easier to maintain and extend. The monolithic architecture has been broken into several classes with clearly separated concerns. 

All PDF generation is now decoupled from the environment. This is in preparation to add a a command line or remote API component in the future.

MPDF has been updated to version 8.3.1

Service classes:

  - **PdfExportService.php**: PDF generation and delivery service
    - Extracted PDF generation logic from action plugin
    - Handles PDF output, caching, and HTTP response
    - Clean separation between event handling and business logic
  - **Config.php**: Centralized configuration management
    - Encapsulates plugin configuration and input variables
    - Generates mPDF configuration arrays
    - Provides clean API for accessing settings
  - **DokuPdf.php**: mPDF wrapper (replaces DokuPDF.class.php)
    - Thin wrapper around mPDF with DokuWiki-specific defaults
    - Custom HttpClient and LocalContentLoader integration
    - has a few todo markers for future improvements in language handling
  - **Writer.php**: Main PDF composition orchestrator
    - Composites Config, Template, DokuPdf, and Styles
    - Handles TOC generation, internal link rewriting, and debug output
    - Manages the overall PDF writing workflow
  - **Template.php**: Template file handling
    - Loads header/footer/cover templates
    - Applies placeholder substitution
  - **Styles.php**: CSS and styling management
    - Loads styles from plugins and templates
    - Handles CSS compilation and processing
  - **Cache.php**: Caching layer
    - Implements DokuWiki cache integration
    - Manages cache validation and retrieval

New collector architecture is responsible to collect the pages to be added to the PDF.

  - **AbstractCollector.php**: Base class for all collectors
    - Common functionality for page collection
    - Checks ACLs
    - extracts title and language info
  - **PageCollector.php**: Single page export
    - No longer relies on global $ID
    - Clean dependency injection via Config
  - **NamespaceCollector.php**: Namespace export
    - Recursive page collection from namespaces
    - provides new event `DW2PDF_NAMESPACEEXPORT_SORT` where plugins could influence the sorting (or even filter out pages)
  - **BookCreatorLiveSelectionCollector.php**: Live BookCreator selections
    - Handles temporary book selections
  - **BookCreatorSavedSelectionCollector.php**: Saved BookCreator selections
    - Handles saved book collections
  - **CollectorFactory.php**: Factory for instantiating collectors
    - Determines correct collector based on request parameters

Media Handling #527

  - **MediaLinkResolver.php**: Media link resolution (replaces old ImageDecorator)
    - Resolves DokuWiki media links to filesystem paths
    - Integration point for custom image processing
  - **HttpClient.php**: Custom HTTP client for mPDF
    - Implements mPDF HttpClient interface
    - Integrates MediaLinkResolver for proper media handling
  - **LocalContentLoader.php**: Custom local content loader
    - Implements mPDF LocalContentLoaderInterface
    - Integrates MediaLinkResolver for proper media handling

Link rewriting is now done outside the (cached) rendering. The plugin requires the DOM and XML extension for this.

A bunch of tests have been added but they are far from comprehensive. **Help in manually testing this branch would be very welcome**

In theory, bookcreator integration should remain unchanged, but has not been tested. I did however change a lot of the Exception handling regarding the bookcreator integration so this part might behave different.


Issues addressed in this PR: #528 #527 #526 #508 #455 #140 (could be implemented via the new event now)


> This refactoring has been largely financed by the [DokuWiki Business Plugin Partner Program](https://www.cosmocode.de/en/services/wiki/dokuwiki-partnership/). Please consider joining the program to support future work like this.

splitbrain committed:
updated mpdf to 8.2 (requires php 8.0 now)

splitbrain committed:
first start at refactoring

We now have a few classes that separate the concerns

DokuPDF  - inherits from mPDF and doesn't do much anymore except for a
           few internal defaults and overwrites
Config   - encapsulate the configuration (file and input vars) and created
           an mpdf config array
Template - handles loading the template files and applies placeholders
Writer   - composites the above classed and gives access to write
           operations

splitbrain committed:
add toc handling

splitbrain committed:
handle debug writing

splitbrain committed:
remove a few now unused methods

splitbrain committed:
Handle styles

We now have a Styles class that takes care of loading the styles from
plugins etc.

splitbrain committed:
move page collecting into separate classes

splitbrain committed:
remove some moved methods

splitbrain committed:
remove obsolete template event

splitbrain committed:
more refactoring

The basic handling of page collecting, caching and rendering is now
defined.

There is still some ugly dependency in the renderer and the
ImageProcessor loading is not reimplemented yet.

After the above a first attempt to actually run the code can be made.

splitbrain committed:
decouple renderer

This starts to approach #528 and #526

splitbrain committed:
removed more obsolete methods from action plugin

splitbrain committed:
fixed issues from first real run

splitbrain committed:
extract $INPUT dependencies out of Collector classes

This should make tests and a future command line interface easier to
implement.

splitbrain committed:
reimplement ImageProcessor differently

As suggested in mpdf/mpdf#2145 the previous decorator functionality was
implemented via custom HTTPClient and LocalContent loaders.

splitbrain committed:
updated tests for Media resolving

splitbrain committed:
update general test

splitbrain committed:
removed unneeded custom exception

splitbrain committed:
rewrite internal links. implements #526

splitbrain committed:
address some of the remaining FIXMEs

splitbrain committed:
fix additional blank page at the end

splitbrain committed:
fixed remaining sort test

splitbrain committed:
have collectors check for page existance

splitbrain committed:
for exporting single pages do not rely on $ID

We want the ID to be passed via Config as well.

splitbrain committed:
Added tests

These tests have been initially LLM written, I went through each of them
to make sure they make sense and fixed what I thought needed fixing.

However coverage is still not great and should be improved in the
future.

splitbrain committed:
rector and codesnifffer fixes

splitbrain committed:
remove obsolete book chapter handling in action

This is now handled within the renderer itself

splitbrain committed:
extracted PDF generation and sending out of action

Now only event handling happens in the action handler

Klap-in commented on src/Config.php:
```diff
@@ -0,0 +1,188 @@
+    /**
+     * Get all configuration for mpdf as array
+     *
+     * Note: wrtiting direction needs to be set separately
```
writing

Klap-in commented on action.php:
```diff
@@ -423,20 +428,13 @@ protected function generatePDF($cachefile, $event)
-        require_once(__DIR__ . "/DokuPDF.class.php");
+        $config = new Config($this->conf, $this->getDocumentLanguage($this->list[0]));
+        $mpdf = new DokuPDF($config);
+        $template = new Template($this->getConf('template'), $this->getConf('qrcodescale'));
```
Only template from config setting. I intend to make it modifiable via BookCreator.

Klap-in commented on src/Writer.php:
```diff
@@ -0,0 +1,215 @@
+
+            if ($header) {
+                $this->mpdf->SetHTMLHeader($header, 'O');
+                $this->mpdf->SetHTMLHeader($header, 'E');
```
I did not expect both O and E need to be set for a first page. Does this mean also a second page in the document is always affected? (not tested yet)
(the old code used named html with pseudo css, difficult to see quickly what happened before)

splitbrain replied:
good question. I guess the first page is always odd? OTOH the first page might be pushed to an even page if there's a TOC before? I think it's okay to set both - mpdf will pick what it needs.

Klap-in commented:
Doc change: seems debug output can now directly be requested with `debug` url parameter, it is not needed to enable `$conf['allowdebug'] ` first before using `&debughtml=text|html`

Klap-in reviewed (commented):
Some first comments/questions. Some are also a bit note to self as I have not tested anything.

splitbrain commented on src/Config.php:
```diff
@@ -57,6 +55,8 @@ public function loadPluginConfig(array $conf)
             $this->useStyles = array_map('trim', $this->useStyles);
             $this->useStyles = array_filter($this->useStyles);
         }
+        if (isset($conf['watermark'])) $this->watermark = $conf['watermark']; // FIXME currently not in default config
```
Yeah I guess you rarely would set a default. But I could imagine setting it to "CONFIDENTIAL" in some wikis to make sure people are careful with printouts...

splitbrain commented on src/AbstractCollector.php:
```diff
@@ -27,7 +27,12 @@ public function __construct(?int $rev = null, ?int $at = null)
+        $this->pages = array_filter(
+            array_map('cleanID', $this->collect()),
+            fn($page) => auth_quickaclcheck($page) >= AUTH_READ
+        );
```
I believe `book_skipforbiddenpages` is part of the mentioned changed exception handling...

Collectors will always obey ACLs thanks to above code. I think before a book export that contained ACL protected pages did return to the book creator page with an error unless skipforbiddenpages was set. Now it basically behaves like skipforbiddenpages is always set, because above code will skip those protected pages by default.

splitbrain committed:
some minor fixes based on PR feedback

splitbrain commented on src/Writer.php:
```diff
@@ -0,0 +1,215 @@
+
+        $this->mpdf->WriteHTML($html, HTMLParserMode::HTML_BODY, false, false);
+
+        $this->breakAfterMe();
```
I think you're right. I added the call to applyHeaderFooters

splitbrain committed:
Use attributes in config class to denote how to initialize them

This should make it easier to add new configuration and see with a
glance how it is initialized.

splitbrain committed:
make PdfExportService reusable in debug mode

This refactors the service so that tests can easily access the produced
debug HTML

splitbrain committed:
introduce a first test for rendering - testing numbered headlines

The test passes, but I am not 100% sure the behaviour is correct yet.
Static variables in the renderer might be a problem. Not tested with
multiple documents yet.

Currently render options need to be set via global conf and are not
passed from the Config object.

splitbrain committed:
Overhaul chapter/header counting in renderer

Using static variables makes it impossible to automatically test this
correctly.

The PDF renderer is now a singleton. The Writer class initializes it
anew for each export, but during the export, the same renderer is
reused.

splitbrain committed:
pass the Config object into the renderer

This streamlines the configuration to a single point of entry. The
renderer was the last bit not adhering to that.

splitbrain committed:
Some more tests testing rendering functionality

Again mostly LLM generated but manually checked and corrected.

splitbrain committed:
fix config attrbutes and extend tests

splitbrain committed:
update MPDF downgrade log/psr

mpdf v8.2.6 => v8.2.7
log 3.0.2 => 1.1.4

Using the old log version avoids conflicts with the smtp plugin for now.

However we need a proper solution to this problem in the future
(probably in core).

A PR to be able to use any log version with the smtp plugin has been
submitted: https://github.com/txthinking/Mailer/pull/36

We probably should introduce a log/psr:3 dependency in core when
upgrading to PHP8 (and provide a PSR compatible wrapper around our
logger)

splitbrain committed:
update deleted.files

Note that i accidentally ran and pushed this on master as well. it's
interesting that the issue seemingly did exist on master.

splitbrain committed:
added psr/log

This should have been added in the last commit
837fcadd5edb939d2ba3977f363c3478f965471b but was forgotten.

annda committed:
Use caching settings

annda committed:
Rename DokuPdf class to avoid accidental deletion based on deleted.files

deleted.files includes DokuPDF already, and on some systems the extension manager might delete DokuPdf as well

annda committed:
Update Mpdf to development version which allows custom AssetFetchers

annda committed:
Fix media paths by supplying custom AssetFetcher to Mpdf

annda committed:
Cleanup custom AssetFetcher

splitbrain committed:
Merge pull request #535 from annda/refactor

Use latest mpdf and custom AssetFetcher to fix media paths

annda committed:
Fix revision handling

annda committed:
Fix failing tests

splitbrain committed:
Merge pull request #536 from annda/refactor

Fix revision handling

annda committed:
Set up book title to be available for placeholder replacement

annda committed:
Use safer array access

splitbrain committed:
Merge pull request #537 from annda/refactor

Set up book title to be available for placeholder replacement

splitbrain committed:
Merge branch 'master' into refactor

* master:
  Translation update (sk)
  Version upped
  updated deleted.files
  Translation update (zh-tw)

splitbrain committed:
set default smiley height

Klap-in commented on src/PdfExportService.php:
```diff
@@ -0,0 +1,169 @@
+
+        $writer->startDocument($this->collector->getTitle());
+        // initial context for placeholder replacements, before any pages are loaded
+        $template->setContext($this->collector, $this->collector->getPages()[0], null);
```
collector->getPages() can return an empty array (e.g. NamespaceCollector collecting a not existing namespace or PageCollector for non-existing page, not everywhere that output is considered. Is it useful to continue here? I guess not. What should be the output/feedback for an empty selection?

```
2026-07-19 13:19:41
E_WARNING: Undefined array key 0
/home/user/Sites/dokuwiki/dokuwiki/lib/plugins/dw2pdf/src/PdfExportService.php(153)
#0 /home/user/Sites/dokuwiki/dokuwiki/lib/plugins/dw2pdf/src/PdfExportService.php(153): dokuwiki\ErrorHandler::errorHandler()
#1 /home/user/Sites/dokuwiki/dokuwiki/lib/plugins/dw2pdf/src/PdfExportService.php(112): dokuwiki\plugin\dw2pdf\src\PdfExportService->renderDocument()
#2 /home/user/Sites/dokuwiki/dokuwiki/lib/plugins/dw2pdf/src/PdfExportService.php(58): dokuwiki\plugin\dw2pdf\src\PdfExportService->buildDocument()
#3 /home/user/Sites/dokuwiki/dokuwiki/lib/plugins/dw2pdf/action.php(68): dokuwiki\plugin\dw2pdf\src\PdfExportService->getPdf()
#4 /home/user/Sites/dokuwiki/dokuwiki/inc/Extension/EventHandler.php(80): action_plugin_dw2pdf->convert()
#5 /home/user/Sites/dokuwiki/dokuwiki/inc/Extension/Event.php(75): dokuwiki\Extension\EventHandler->process_event()
#6 /home/user/Sites/dokuwiki/dokuwiki/inc/ActionRouter.php(83): dokuwiki\Extension\Event->advise_before()
#7 /home/user/Sites/dokuwiki/dokuwiki/inc/ActionRouter.php(49): dokuwiki\ActionRouter->setupAction()
#8 /home/user/Sites/dokuwiki/dokuwiki/inc/ActionRouter.php(62): dokuwiki\ActionRouter->__construct()
#9 /home/user/Sites/dokuwiki/dokuwiki/inc/actions.php(19): dokuwiki\ActionRouter::getInstance()
#10 /home/user/Sites/dokuwiki/dokuwiki/doku.php(131): act_dispatch()
#11 {main}

2026-07-19 13:20:46
TypeError: dokuwiki\plugin\dw2pdf\src\Template::setContext(): Argument #2 ($id) must be of type string, null given, called in /home/user/Sites/dokuwiki/dokuwiki/lib/plugins/dw2pdf/src/PdfExportService.php on line 153
/home/user/Sites/dokuwiki/dokuwiki/lib/plugins/dw2pdf/src/Template.php(60)
#0 /home/user/Sites/dokuwiki/dokuwiki/lib/plugins/dw2pdf/src/PdfExportService.php(153): dokuwiki\plugin\dw2pdf\src\Template->setContext()
#1 /home/user/Sites/dokuwiki/dokuwiki/lib/plugins/dw2pdf/src/PdfExportService.php(112): dokuwiki\plugin\dw2pdf\src\PdfExportService->renderDocument()
#2 /home/user/Sites/dokuwiki/dokuwiki/lib/plugins/dw2pdf/src/PdfExportService.php(58): dokuwiki\plugin\dw2pdf\src\PdfExportService->buildDocument()
#3 /home/user/Sites/dokuwiki/dokuwiki/lib/plugins/dw2pdf/action.php(68): dokuwiki\plugin\dw2pdf\src\PdfExportService->getPdf()
#4 /home/user/Sites/dokuwiki/dokuwiki/inc/Extension/EventHandler.php(80): action_plugin_dw2pdf->convert()
#5 /home/user/Sites/dokuwiki/dokuwiki/inc/Extension/Event.php(75): dokuwiki\Extension\EventHandler->process_event()
#6 /home/user/Sites/dokuwiki/dokuwiki/inc/ActionRouter.php(83): dokuwiki\Extension\Event->advise_before()
#7 /home/user/Sites/dokuwiki/dokuwiki/inc/ActionRouter.php(49): dokuwiki\ActionRouter->setupAction()
#8 /home/user/Sites/dokuwiki/dokuwiki/inc/ActionRouter.php(62): dokuwiki\ActionRouter->__construct()
#9 /home/user/Sites/dokuwiki/dokuwiki/inc/actions.php(19): dokuwiki\ActionRouter::getInstance()
#10 /home/user/Sites/dokuwiki/dokuwiki/doku.php(131): act_dispatch()
#11 {main}
```

splitbrain replied:
fixed

Klap-in commented on src/NamespaceCollector.php:
```diff
@@ -0,0 +1,198 @@
+
+        // check namespace exists
+        $nsdir = dirname(wikiFN($this->namespace . ':dummy'));
+        if (!@is_dir($nsdir)) throw new \Exception('needns');
```
This before ` throw new Exception($this->getLang('needns'));` why no real message?

Klap-in commented on src/NamespaceCollector.php:
```diff
@@ -0,0 +1,198 @@
+        try {
+            $this->initVars();
+        } catch (\Exception $e) {
+            return [];
```
needns Exception is directly catched and returns only empty array, why no message to user?

Klap-in commented on src/PageCollector.php:
```diff
@@ -0,0 +1,22 @@
+
+        // no export for non existing page
+        if (!page_exists($exportID, $this->rev)) {
+            return [];
```
no msg() for feedback to user anymore?

Klap-in commented on src/Config.php:
```diff
@@ -0,0 +1,462 @@
+    {
+        global $INPUT, $ID;
+
+        if ($ID) $this->exportId = $ID; // default exportId to current page ID
```
```suggestion
        if (!blank($ID)) $this->exportId = $ID; // default exportId to current page ID
```
Page 0 (zero) is a valid name: 0?do=export_pdf

splitbrain committed:
add blank line after file header in exportlink.php

splitbrain committed:
fix export of pages named 0

The current page id was only applied as the export default when truthy, so a
page literally named "0" was treated as if no page was given. Use blank() so
"0" is accepted as a valid id.

splitbrain committed:
give user feedback for empty or invalid PDF exports

Collectors used to return an empty page list for a missing page, a missing
namespace or an empty selection, which then failed silently or crashed when the
first page was accessed. They now throw a translated ExportException instead,
which the action handler turns into feedback: a flash message and redirect for
regular navigations, or an HTTP error with the message in the body for the
background download BookCreator uses. Dynamic message parts are HTML escaped
because that body is injected into the page via a jQuery html() sink.

splitbrain committed:
skip forbidden pages and distinguish them from empty selections

Collected pages the user may not read are dropped from the export and remembered
separately, so a selection where every page is forbidden can be told apart from
one where nothing was selected and reports "no read access" instead of "nothing
selected". The forbidden message no longer lists the page ids or references the
removed 'Skip Forbidden Pages' option; its outdated translations are dropped so
they get re-translated.

splitbrain committed:
require allowdebug for debug output requested via URL

The debug URL parameter dumped the raw mPDF HTML and bypassed the cache for any
visitor. It is now honored from the URL only when core allowdebug is enabled;
the config/programmatic path stays trusted.

splitbrain committed:
rector fixes

splitbrain committed:
updated dependencies

mpdf is back to a release tag (my changes have been merged). minimum PHP
version is 8.2 now, just like DokuWiki.
